0004-ledger-plugin-interface/0004-ledger-plugin-interface.md

```diff
@@ -199,12 +208,19 @@ Submit a fulfillment to a ledger. Plugin must be connected, otherwise the promis
 
 **TODO**: Define what the message format is.  Plugin must be connected, otherwise the promise should reject.
 
+#### rejectIncomingTransfer
+<code>ledgerPlugin.rejectIncomingTransfer( **transferId**:String ) ⇒ Promise.&lt;null></code>
+
+Reject an incoming transfer that is held pending the fulfillment of its `executionCondition` before the `expiresAt` time.
+
```

Comment:
> The protocol doesn't allow or disallow any functionality ledgers (and corresponding plugins) choose to implement. 

I disagree. A protocol, by definition, MUST define what implementations of the protocol are allowed and not allowed to do. That's what a protocol is, a set of rules that everyone must follow to be interoperable.

> This interface specifies the functionality necessary for clients and connectors to use plugins interchangeably for sending and receiving ILP payments.

I think this is the crux of the misunderstandings around this plugin interface and the protocol itself. The protocol defines the required behaviour of ledgers and connectors however the ledger plugin is a representation of the ledger interface running in the connector's processing space.

The design of the ledger interface **for the purpose of defining the requirements of a ledger to be conformant to ILP** should be a ledger API (even an abstract one) that is assumed to be running in the ledger's processing space.

Once you have this interface defined you can define a way this can be implemented as a plugin that runs in the connector's processing space instead.

By doing it as a plug-in first we are making design decisions or inferring things about the protocol that are not correct.

For example, you have said that _the "authorized credits" feature is not part of this interface_ but that's an implementation decision. The protocol needs to define how a ledger and connector interact when a transfer is received by a connector that has no condition and the connector wishes to reject it. 

The reasons for rejecting the transfer (eg: it's not authorized) are irrelevant, those are business rules which are implementation specific. The rules of the protocol either need to prevent this from happening (all transfers must have a condition) or define the behavior of the participants when it does (receivers can reject an incoming transfer even if it has no condition).

> Note that ledger plugin implementations MAY emit more events and provide more functionality than specified here. However, we should define the interface such that the vast majority will implement all of the minimum functionality described here.

No, a protocol needs to define the behavior for all cases not just the majority. If this interface is supposed to be an implicit definition of ILP then it must define the necessary events and methods used by all participants for all cases.

If the purpose of this interface is not to cover all cases for an ILP transfer then we need to be explicit about that and have somewhere else that this behaviour is documented.

> Send it back or keep it.

How? How does a connector that can't route a payment send it back and notify the sender that it is linked to the original payment? The protocol needs to define this behavior, if not in this interface then we need to do it somewhere else.